edb/schema/objects.py

```diff
@@ -1329,24 +1252,174 @@ def as_shell(self, schema: s_schema.Schema) -> ObjectShell:
+                None,
+                self,
+                delta=delta,
+                refdict=refdict,
+                context=context,
+                old_schema=None,
+                new_schema=schema,
+            )
+
+        return delta
+
+    def as_alter_delta(
+        self: Object_T,
+        other: Object_T,
+        *,
+        self_schema: s_schema.Schema,
+        other_schema: s_schema.Schema,
+        context: ComparisonContext,
+    ) -> sd.ObjectCommand[Object_T]:
+        from . import delta as sd
+
+        cls = type(self)
+        delta = self.init_delta_command(
+            self_schema,
+            sd.AlterObject,
+            canonical=True,
+        )
+        cls.delta_properties(
+            delta,
+            self,
+            other,
+            context=context,
+            old_schema=self_schema,
+            new_schema=other_schema,
+        )
+
+        for refdict in cls.get_refdicts():
+            cls._delta_refdict(
+                self,
+                other,
+                delta=delta,
+                refdict=refdict,
+                context=context,
+                old_schema=self_schema,
+                new_schema=other_schema,
+            )
+
+        return delta
+
+    def as_delete_delta(
+        self: Object_T,
+        *,
+        schema: s_schema.Schema,
+        context: ComparisonContext,
+    ) -> sd.ObjectCommand[Object_T]:
+        from . import delta as sd
+
+        cls = type(self)
+        delta = self.init_delta_command(
+            schema,
+            sd.DeleteObject,
+            canonical=True,
+        )
+        cls.delta_properties(
+            delta,
+            self,
+            None,
+            context=context,
+            old_schema=schema,
+            new_schema=None,
+        )
+
+        for refdict in cls.get_refdicts():
+            cls._delta_refdict(
+                self,
+                None,
+                delta=delta,
+                refdict=refdict,
+                context=context,
+                old_schema=schema,
+                new_schema=None,
+            )
+
+        return delta
+
     @classmethod
     def delta_properties(
-        cls,
-        delta: sd.ObjectCommand[Object],
-        old: Optional[Object],
-        new: Optional[Object],
+        cls: Type[Object_T],
+        delta: sd.ObjectCommand[Object_T],
+        old: Optional[Object_T],
+        new: Optional[Object_T],
         *,
         context: ComparisonContext,
         old_schema: Optional[s_schema.Schema],
-        new_schema: s_schema.Schema,
+        new_schema: Optional[s_schema.Schema],
     ) -> None:
         ff = cls.get_fields(sorted=True).items()
         fields = {fn: f for fn, f in ff
                   if f.simpledelta and not f.ephemeral and f.introspectable}
 
         if old and new:
-            if old_schema is None:
-                raise ValueError("`old` provided but `old_schema is None")
+            assert \
```

Comment:
nit: IMO a regular `if ... raise AssertionError` is better in cases where so many `\` are used.

Reply:
I'm still holding hope that we might be able to disable assertion checks at some point given the amount of them we've been adding to make mypy happy.